gpii/node_modules/flowManager/src/PSPChannel.js

```diff
@@ -182,30 +308,31 @@ gpii.pspChannel.emitSettingControl = function (settingControls, schema, prefVal,
  *     - {Object} preferences - A filtered skeleton of the user's preferences document, just containing the names of preferences sets (`contexts`).
  */
 gpii.pspChannel.sessionToPSP = function (model, transformSpec) {
-    var schemas = transformSpec.ontologyMetadata.flat;
+    var that = transformSpec.pspChannel;
+    // populate the default setting controls with schema defaults at the first time
+    if (!that.defaultSettingControls) {
+        that.defaultSettingControlsPromise.then(function (defaultSettingControls) {
+            that.defaultSettingControls = defaultSettingControls;
```

Comment:
Horrible risk of a race here since this value may be consumed on line 335 before it is ready. Also, this is the time that it is most crucial that the payload is ready since the most important use case is that the GPII app's indicators have the correct state on startup.
What I suggest is - 
i) Create a further event for the resolution of the default setting controls
ii) Rename and repurpose the existing "readyToReset" event as a new event "flowManagerReady" which only fires once all outstanding I/O has completed
iii) Delay the firing of noUserLoggedIn until this new event flowManagerReady has fired, rather than waiting immediately on {kettle.server}.events.onListen
iv) Add a "createOnEvent" to the pspChannel to delay its construction until flowManagerReady has 
This will then be consistent with the strategy used in pull https://github.com/GPII/gpii-app/pull/117/files for GPII-3818 "indicative fix" which I believe that @javihernandez is still rolling into current installers, since this adds a delay waiting for "resetAtStartSuccess" which then cannot occur before all of the FlowManager's I/O is complete.
Also please add a comment indicating that all of this machinery should be removed once we can update to the FLUID-6148 framework and we can put all of this I/O into the "resources" block of a fluid.resourceLoader.

Reply:
All suggestions are implemented except the fourth about adding a "createOnEvent" to the pspChannel to delay its construction. This point is difficult or even impossible to implement without a relatively big system refactoring because a number of other components such as lifecycleManager is referencing to "{pspChannel}" while "{lifecycleManager}" is referenced by pspChannel and flowManager. Adding a "createOnEvent" to pspChannel will cause an "implementation error" at the system startup. 

My workaround is to continue to generate "defaultSettingControls" at [handling the first noUser key in request](https://github.com/cindyli/universal/blob/GPII-3828/gpii/node_modules/flowManager/src/PSPChannel.js#L312-L314) rather than using an expander at the construction of pspChannel.